MHU: Add MHU driver support for SSE-200 subsystem

karl-zh opened the pull request:
MHU (Message Handling Unit) enable software to raise interrupts to
the processor cores.

This patch enable MHU driver on Musca A which has 2 MHU elements. It helps to communicate between 2 cores.

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

Commit:
drivers: ipm: mhu: Add MHU driver for V2M Musca

MHU (Message Handling Unit) enables software to raise interrupts to
the processor cores. It is enabled in SSE 200 subsystems.

This patch aims to implement inter processor communication.

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

MaureenHelm reviewed (changes requested):
Please look at implementing the IPM interface (`include/ipm.h`) rather than introducing a new interface.

galak commented:
> Please look at implementing the IPM interface (`include/ipm.h`) rather than introducing a new interface.

Agreed, please use the IPM interface for this driver.

Also, do we need some code and support for bringing up the second core in Zephyr, and running Zephyr on it?

karl-zh commented:
> > Please look at implementing the IPM interface (`include/ipm.h`) rather than introducing a new interface.
> 
> Agreed, please use the IPM interface for this driver.
> 
> Also, do we need some code and support for bringing up the second core in Zephyr, and running Zephyr on it?

Okay, updated to ipm interface. I was thought that the MHU has no data transfer compare to ipm.
Yes, it needs some change to wake up another core on musca, I will have a sample about use MHU on Musca.

Commit:
sample: mhu: IPM MHU dual core on V2M Musca

Sample walk through:
    1. CPU 0 will wake up CPU 1 after initialization
    2. CPU 1 will send to CPU 0 an interrupt over MHU0
    3. CPU 0 return the same to CPU 1 when received MHU0 interrupt
    4. Test done when CPU 1 received MHU0 interrupt

The wake up second core and private core ID are soc specific.

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

stanislav-poboril commented on drivers/ipm/ipm_mhu.c:
```diff
@@ -0,0 +1,245 @@
+#define DEV_DATA(dev) \
+	((struct ipm_mhu_data *)(dev)->driver_data)
+#define IPM_MHU_REGS(dev) \
+	((volatile struct ipm_mhu_reg_map_t  *)(DEV_CFG(dev))->base)
```
two spaces between ipm_mhu_reg_map_t and *

stanislav-poboril commented on drivers/ipm/ipm_mhu.h:
```diff
@@ -0,0 +1,74 @@
+#define SSE_200_CPU_ID_UNIT_OFFSET	((0x1F000UL))
+
+/* SSE 200 MHU register map structure */
+struct ipm_mhu_reg_map_t {
```
Are these definitions somewhere in SoC's vendor SDK? If so, the vendor headers could be put into ext/hal/... and included and reused here (assuming there are no license issues).

galak replied:
The ARM IP has followed this model of not having a SDK or headers, so its fine that we define the register map here.

stanislav-poboril replied:
Ok. Btw I think checkpatch.pl ignores the ext folder?

stanislav-poboril reviewed (approved):
Looks good, I have some comments to the commit message:

Maybe it should start with "drivers/ipm: "?
Shouldn't the word "enables" be here: MHU (Message Handling Unit) enable software
And "It enabled in SSE 200 subsystems." seems to miss "is" perhaps.

galak commented on drivers/ipm/Kconfig:
```diff
@@ -75,3 +75,21 @@ config IPM_IMX_MAX_ID_VAL
+	help
+	  Driver for SSE 200 MHU (Message Handling Unit)
+
+config IPM_MHU0
```
Do we need instance defines, can we get this from DT?

`DT_ARM_MHU_0` or `DT_ARM_MHU_1`

karl-zh replied:
DT_ARM_MHU_0 and DT_ARM_MHU_1 are always defined after mhu added into dts file, are they?
The config here provide a choice which MHU device should add as in this sub system include 2 MHU.
When enabling MHU device, there will be settings in the prj.conf like:

CONFIG_IPM=y
CONFIG_IPM_MHU=y
CONFIG_IPM_MHU0=y
CONFIG_IPM_MHU1=y

galak replied:
But that choice of which MHU can and should be handled by whats in the DTS.   I don't think we need the Kconfig options here.

galak commented on drivers/ipm/ipm_mhu.c:
```diff
@@ -0,0 +1,245 @@
+#define IPM_MHU_REGS(dev) \
+	((volatile struct ipm_mhu_reg_map_t  *)(DEV_CFG(dev))->base)
+
+#if defined(CONFIG_IPM_MHU0) || defined(CONFIG_IPM_MHU1)
```
This ifdef shouldn't be needed.

karl-zh replied:
CONFIG_IPM=y
CONFIG_IPM_MHU=y
#CONFIG_IPM_MHU0=y
#CONFIG_IPM_MHU1=y

If config file above, there are warnings about some symbols declared but not used. Do we need to consider this scenario?

galak commented on drivers/ipm/ipm_mhu.c:
```diff
@@ -0,0 +1,245 @@
+
+	p_mhu_dev_base = (volatile u32_t *)IPM_MHU_REGS(d);
+
+	p_cpu_id = (volatile u32_t *)(((u32_t)p_mhu_dev_base & 0xF0000000) +
```
Can we add a #define for the magic number 0xF0000000

galak commented on boards/arm/v2m_musca/v2m_musca-common.dtsi:
```diff
@@ -22,6 +22,18 @@ dtimer0: dtimer@2000 {
 	interrupts = <5 3>;
 };
 
+mhu0: mhu@3000 {
```
We should add `label` properties, its the current way we use in Zephyr to kinda map device_get_binding() and DTS.

galak commented on dts/bindings/mhu/arm,mhu.yaml:
```diff
@@ -0,0 +1,27 @@
+      description: compatible strings
+      constraint: "arm,mhu"
+      generation: define
+
```
add a label property here

karl-zh replied:
Added the MHU label, and use it in the code, commit 51ac287cffd90214aef115b649a71fda6873f669.
Thanks

galak reviewed (changes requested):
Looks good, some minor review comments in a few places.

dbkinder commented on samples/subsys/ipc/ipm_mhu_dual_core/README.rst:
```diff
@@ -0,0 +1,48 @@
+.. _ipm_mhu_dual_core:
+
+MHU Dual Core
+###########
```
The heading underline must be at least as long as the title, so add a couple more ``##``

dbkinder commented on samples/subsys/ipc/ipm_mhu_dual_core/README.rst:
```diff
@@ -0,0 +1,48 @@
+
+Overview
+********
+A simple dual core example that can be used with arm Musca A1
```
After reading this description I have no idea what this sample does or what it's for (but you do mention some facts in the commit message that should be added here.

How about:
```
An MHU (Message Handling Unit) enables software to raise interrupts to
the processor cores.  This sample is a simple dual-core example for a
Musca A1 board that has two MHU units.
```

And then can you add some more explanation of what the sample does?

karl-zh replied:
Thanks, I have updated the documents in latest commit.

dbkinder commented on samples/subsys/ipc/ipm_mhu_dual_core/README.rst:
```diff
@@ -0,0 +1,53 @@
+the processor cores. This sample is a simple dual-core example for a
+Musca A1 board that has two MHU units. This sample only test MHU0, the
+steps are:
+1. CPU 0 will wake up CPU 1 after initialization
```
add a blank line before the first item in a list

dbkinder commented on samples/subsys/ipc/ipm_mhu_dual_core/README.rst:
```diff
@@ -0,0 +1,53 @@
+3. CPU 0 return the same to CPU 1 when received MHU0 interrupt
+4. Test done when CPU 1 received MHU0 interrupt
+
+* single thread
```
Not sure what this bullet point is for?  Delete it?

dbkinder reviewed (changes requested):
getting closer on the docs...

karl-zh commented:
> getting closer on the docs...

Thanks, updated now.

karl-zh commented on drivers/ipm/ipm_mhu.c:
```diff
@@ -0,0 +1,246 @@
+			  const void *data, int size)
+{
+	ARG_UNUSED(wait);
+	u32_t set_val = *(u32_t *)data;
```
The register value should not from the param.
Suggest to use a fixed number instead:

	ARG_UNUSED(data);
	u32_t set_val = 0x01;

galak commented on samples/subsys/ipc/ipm_mhu_dual_core/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+tests:
+  test:
+    tags: ipm
+    platform_whitelist: arm Musca A1 board
```
This doesn't look right.

MaureenHelm commented on samples/subsys/ipc/ipm_mhu_dual_core/src/main.c:
```diff
@@ -0,0 +1,100 @@
+
+struct device *mhu0;
+
+static void wakeup_cpu1(void)
```
Why isn't this in the soc or board-level support?

karl-zh replied:
@galak Hi Kumar,
What is your opinion on that? The driver would be in ext/hal/xxx folder for each platform, right? Current Zephyr seems no arm folder in ext/hal/.

MaureenHelm replied:
Why would it be in `ext/`? I think it belongs in `soc/arm/arm/musca_a`

MaureenHelm commented on drivers/ipm/ipm_mhu.c:
```diff
@@ -0,0 +1,235 @@
+
+static void ipm_mhu_irq_config_func_0(struct device *d);
+
+static struct ipm_mhu_device_config ipm_mhu_cfg_0 = {
```
This can be `const`

MaureenHelm commented on drivers/ipm/ipm_mhu.c:
```diff
@@ -0,0 +1,235 @@
+
+static void ipm_mhu_irq_config_func_1(struct device *d);
+
+static struct ipm_mhu_device_config ipm_mhu_cfg_1 = {
```
This can also be `const`

MaureenHelm commented on dts/bindings/mhu/arm,mhu.yaml:
```diff
@@ -0,0 +1,33 @@
+---
```
Add a license/copyright header

MaureenHelm commented on samples/subsys/ipc/ipm_mhu_dual_core/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+cmake_minimum_required(VERSION 3.13.1)
```
Add a license/copyright header

MaureenHelm commented on drivers/ipm/ipm_mhu.c:
```diff
@@ -0,0 +1,235 @@
+
+static void ipm_mhu_irq_config_func_0(struct device *d);
+
+const struct ipm_mhu_device_config ipm_mhu_cfg_0 = {
```
This should be `static` AND `const`

MaureenHelm commented on drivers/ipm/ipm_mhu.c:
```diff
@@ -0,0 +1,235 @@
+
+static void ipm_mhu_irq_config_func_1(struct device *d);
+
+const struct ipm_mhu_device_config ipm_mhu_cfg_1 = {
```
`static const`

MaureenHelm commented on samples/subsys/ipc/ipm_mhu_dual_core/src/main.c:
```diff
@@ -0,0 +1,74 @@
+
+extern void wakeup_cpu1(void);
+
+extern u32_t sse_200_platform_get_cpu_id(void);
```
Move these externs to `soc.h`

karl-zh replied:
Fixed.
